File: frontend/src/app/core/services/payment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  PaymentTransaction, RecordPaymentRequest,
  ReversalRequest, ScheduleAdjustmentRequest
} from '../models/payment.model';
import { ScheduleVersion } from '../models/loan.model';

@Injectable({ providedIn: 'root' })
export class PaymentService {
  constructor(private http: HttpClient) {}

  getSchedule(loanId: string): Observable<ScheduleVersion> {
    return this.http.get<ScheduleVersion>(`/api/v1/loans/${loanId}/schedule`);
  }

  getPayments(loanId: string): Observable<{ items: PaymentTransaction[] }> {
    return this.http.get<{ items: PaymentTransaction[] }>(`/api/v1/loans/${loanId}/payments`);
  }

  recordPayment(loanId: string, data: RecordPaymentRequest): Observable<PaymentTransaction> {
    const idempotencyKey = crypto.randomUUID();
    return this.http.post<PaymentTransaction>(
      `/api/v1/loans/${loanId}/payments`, data,
      { headers: { 'Idempotency-Key': idempotencyKey } }
    );
  }

  reversePayment(paymentId: string, data: ReversalRequest): Observable<PaymentTransaction> {
    const idempotencyKey = crypto.randomUUID();
    return this.http.post<PaymentTransaction>(
      `/api/v1/payments/${paymentId}/reversals`, data,
      { headers: { 'Idempotency-Key': idempotencyKey } }
    );
  }

  reschedule(loanId: string, data: ScheduleAdjustmentRequest): Observable<void> {
    return this.http.post<void>(`/api/v1/loans/${loanId}/schedule-adjustments/reschedule`, data);
  }

  pause(loanId: string, data: ScheduleAdjustmentRequest): Observable<void> {
    return this.http.post<void>(`/api/v1/loans/${loanId}/schedule-adjustments/pause`, data);
  }

  getHistory(loanId: string): Observable<{ items: unknown[] }> {
    return this.http.get<{ items: unknown[] }>(`/api/v1/loans/${loanId}/history`);
  }
}

File: frontend/Services/PaymentService.cs
using System.Net.Http.Json;
using LendQ.Client.Models;

namespace LendQ.Client.Services;

public class PaymentService : IPaymentService
{
    private readonly HttpClient _http;

    public PaymentService(HttpClient http)
    {
        _http = http;
    }

    public event Action? OnChange;

    public async Task<ScheduleVersion> GetScheduleAsync(string loanId)
    {
        return await _http.GetFromJsonAsync<ScheduleVersion>($"loans/{loanId}/schedule")
            ?? throw new InvalidOperationException("Schedule not found.");
    }

    public async Task<List<PaymentTransaction>> GetPaymentsAsync(string loanId)
    {
        return await _http.GetFromJsonAsync<List<PaymentTransaction>>($"loans/{loanId}/payments")
            ?? new List<PaymentTransaction>();
    }

    public async Task<PaymentTransaction> RecordPaymentAsync(string loanId, RecordPaymentFormModel model)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, $"loans/{loanId}/payments");
        request.Headers.Add("Idempotency-Key", Guid.NewGuid().ToString());
        request.Content = JsonContent.Create(model);

        var response = await _http.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var payment = await response.Content.ReadFromJsonAsync<PaymentTransaction>()
            ?? throw new InvalidOperationException("Record payment response was null.");
        OnChange?.Invoke();
        return payment;
    }

    public async Task<PaymentTransaction> ReversePaymentAsync(string paymentId, string reason)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, $"payments/{paymentId}/reverse");
        request.Headers.Add("Idempotency-Key", Guid.NewGuid().ToString());
        request.Content = JsonContent.Create(new { reason });

        var response = await _http.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var payment = await response.Content.ReadFromJsonAsync<PaymentTransaction>()
            ?? throw new InvalidOperationException("Reverse payment response was null.");
        OnChange?.Invoke();
        return payment;
    }

    public async Task RescheduleAsync(string loanId, ScheduleAdjustmentFormModel model)
    {
        var response = await _http.PostAsJsonAsync($"loans/{loanId}/schedule/reschedule", model);
        response.EnsureSuccessStatusCode();
        OnChange?.Invoke();
    }

    public async Task PauseAsync(string loanId, ScheduleAdjustmentFormModel model)
    {
        var response = await _http.PostAsJsonAsync($"loans/{loanId}/schedule/pause", model);
        response.EnsureSuccessStatusCode();
        OnChange?.Invoke();
    }

    public async Task<List<object>> GetHistoryAsync(string loanId)
    {
        return await _http.GetFromJsonAsync<List<object>>($"loans/{loanId}/schedule/history")
            ?? new List<object>();
    }
}
